src/libs/manymouse/linux_evdev.c

```diff
@@ -216,33 +218,48 @@ static int init_mouse(__attribute__((unused)) const char *fname, int fd)
 } /* init_mouse */
 
 
-/* Return a file descriptor if this is really a mouse, -1 otherwise. */
+/* Return 1 if this is really a mouse, 0 otherwise. */
 static int open_if_mouse(const char *fname)
 {
+    assert(fname);
+
+    /* Can we open the file device? */
+    const int fd = open(fname, O_RDONLY | O_NONBLOCK);
+    if (fd == -1)
+        return 0;
+
     struct stat statbuf;
-    int fd;
-    int devmajor, devminor;
+    memset(&statbuf, 0, sizeof(statbuf));
 
-    if (stat(fname, &statbuf) == -1)
+    /* Can we stat info about the device? */
+    if (fstat(fd, &statbuf) == -1) {
+        close(fd);
         return 0;
+    }
 
-    if (S_ISCHR(statbuf.st_mode) == 0)
-        return 0;  /* not a character device... */
+    /* Is it a character device? */
+    if (S_ISCHR(statbuf.st_mode) == 0) {
+        close(fd);
+        return 0;
+    }
 
     /* evdev node ids are major 13, minor 64-96. Is this safe to check? */
-    devmajor = (statbuf.st_rdev & 0xFF00) >> 8;
-    devminor = (statbuf.st_rdev & 0x00FF);
-    if ( (devmajor != 13) || (devminor < 64) || (devminor > 96) )
+    const int devmajor = (statbuf.st_rdev & 0xFF00) >> 8;
+    const int devminor = (statbuf.st_rdev & 0x00FF);
+    if ( (devmajor != 13) || (devminor < 64) || (devminor > 96) ) {
+        close(fd);
         return 0;  /* not an evdev. */
+    }
 
-    if ((fd = open(fname, O_RDONLY | O_NONBLOCK)) == -1)
```

Comment:
Skipping this check is probably not a good idea - it would be safer to check whether we are actually able to open the device. On Linux it is always possible we don't have credentials.

Reply:
Yup! (just moved first thing)

![2022-10-22_14-55](https://user-images.githubusercontent.com/1557255/197363844-c7178243-0c1f-40c2-9880-18f527fc0013.png)

The file-handle is reused if the open was successful. It's the re-use of the handle in the stat-checks that zips up the TOCTOU gap.